README.md
## S3 File Connector

CI Status: ![CI Status](https://github.com/awslabs/s3-file-connector/actions/workflows/rust.yml/badge.svg?branch=main)

## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.

## License

This project is licensed under the Apache-2.0 License.


CHANGELOG.md
# FUSE for Rust - Changelog

## 0.12.0 - 2022-12-13
* Add method to `Session` to unmount non-`Send` `Filesystem`s

## 0.11.1 - 2022-08-24
* Improve an error message when using libfuse2

## 0.11.0 - 2022-03-05
* Add `spawn_mount2()`
* Deprecate `spawn_mount()`

## 0.10.0 - 2022-01-06
* Improve error messages
* Support compiling with musl
* Default `link()` & `symlink()` now return EPERM instead of ENOSYS

## 0.9.1 - 2021-09-07
* `forget` and `batch_forget` no longer require that `AllowRoot` be set

## 0.9.0 - 2021-08-31
* Ensure that `Filesystem::destroy` is always called, when the filesystem is unmounted
* Remove request parameter from `Filesystem::destroy`.
* Make `fuse_forget_one` public, so that `Filesystem::batch_forget` can be implemented by users.
* Fix `batch_forget`. Previously, it always received an empty list of inodes.
* Fix `MountOption::AllowRoot`. Previously, using it resulted in a crash.
* Fix `MountOption::AutoUnmount` so that it works when `AllowRoot` and `AllowOther` are both not set.
* Make log messages more verbose (now includes the operation)

## 0.8.0 - 2021-06-11
* Deprecate `mount()`
* Remove `FileAttr.padding`. This field was added by mistake, and does nothing
* Fix crash when receiving an unknown FUSE operation type
* Minor performance optimizations

## 0.7.0 - 2021-01-10
* Support building with MacFuse 4.x on OSX
* Support configuring max_write & max_readahead via `KernelConfig` during `init`
* Support configuring filesystem timestamp granularity via `KernelConfig.set_time_granularity` during `init`
* Support requesting additional capability flags via `KernelConfig.add_capabilities` during `init`

## 0.6.0 - 2020-11-22
* Make `spawn_mount()` safe
* Change `flags` parameter of `create()`, `open()`, `opendir()`, `release()`, `releasedir()` to be signed, so that it matches
  libfuse and the associated constants in libc
* Change `flags` parameter of `setxattr()` to be signed, so that it matches libfuse
* Change `mask` parameter of `access()` to be signed, so that it matches libfuse and the associated constants in libc
* Change lock type parameter of `getlk()` and `setlk()` to be signed, so that it matches libfuse and the associated constants in libc
* Change atime & atime_now and mtime & mtime_now parameters of `setattr()` to make their relationship more obvious
* Add `lock_owner` and file `flags` parameters to `read()` and `write()`
* Add `umask` parameter to `mknod()`, `mkdir()` and `create()`
* Add `KernelConfig` parameter to `init()` to allow `Filesystem` to configure the kernel connection attributes
* Add support for `fallocate()`, `ioctl()`, `copy_file_range()`, and `lseek()`
* Add support for FUSE_BATCH_FORGET
* Add support for FUSE_READDIRPLUS
* Add support for FUSE_RENAME2
* Add FUSE_WRITE_KILL_PRIV flag for `write()`
* Add FUSE_WRITEBACK_CACHE flag
* Add FUSE_NO_OPEN_SUPPORT flag
* Add FUSE_PARALLEL_DIROPS flag
* Add FUSE_HANDLE_KILLPRIV flag
* Add FUSE_POSIX_ACL flag
* Add FUSE_ABORT_ERROR flag
* Add FUSE_NO_OPENDIR_SUPPORT flag
* Add FUSE_CACHE_SYMLINKS flag
* Add FUSE_EXPLICIT_INVAL_DATA flag
* Add FUSE_IOCTL_COMPAT_X32 flag
* Add FOPEN_CACHE_DIR flag
* Add FOPEN_STREAM flag
* Add FUSE_MAX_PAGES flag
* Add max_pages, and time_gran support to init code path (these are not currently configurable)
* Add support for ctime in `setattr()`
* Add support for timestamps before the unix epoch in `getattr()` and `setattr()`

## 0.5.0 - 2020-10-17

* Enable FUSE_BIG_WRITES for ABI >= 7.10
* Add FUSE_AUTO_INVAL_DATA constant
* Add ABI 7.20 to 7.31 feature flags. Support for these are incomplete.
* Add support for building with libfuse3
* Add support for building without libfuse/libfuse3 on Linux (i.e. there's now a pure Rust implementation of all features)
* Add `mount2()` with improved option API

## 0.4.1 - 2020-10-12

* Added new feature `serializable` that will enable serde serialization/deserialization for `FileType`, `FileAttr`

## 0.4.0 - 2020-06-18

* Forked as `fuser` crate, at https://github.com/cberner/fuser
* Add ATIME_NOW and MTIME_NOW support
* Add stubs for ioctl, fallocate, and poll for ABI 7.11

## 0.3.1 - 2017-11-08

* Offsets to `read`, `write` and `readdir` methods are signed integers now (breaking change, sorry)
* Link `libosxfuse` on macOS, `libfuse` on all other systems

## 0.3.0 - 2017-01-06

* Fix extended attribute handling (`getxattr` and `listxattr` methods changed and `ReplyXattr` was added)
* `mount` now also returns a `Result` since it may fail if the session fails to run
* Filenames are now passed as `&OsStr` in the filesystem interface
* Removed publishing of documentation on GitHub pages. Docs are now available on https://docs.rs/fuse
* Add `FileType::Socket`

## 0.2.8 - 2016-07-31

* Documentation of releases is build by CI now and made available at https://zargony.github.io/rust-fuse
* Fix `unmount` on BSD systems
* Simplified `libfuse` detection with `pkg-config`
* `ReplyDirectory::sized` was removed since it was impossible to use it safely

## 0.2.7 - 2015-09-08

* Update to latest Rust stable - no longer needs nightly Rust
* A filesystem implementation doesn't need to be `Send` anymore to be mounted synchronously
* A filesystem implementation doesn't need to be 'static anymore to be mounted asynchronously
* CI tests are covering nightly, beta and stable Rust under OSX and Linux now

## 0.2.6 - 2015-04-23

* Update to latest Rust nightly
* Fix mounting of filesystems as non-root on Linux systems

## 0.2.5 - 2015-03-21

* Update to latest Rust nightly
* `unmount` returns a `Result` now since unmounting may fail internally
* Fix `unmount` on Linux systems
* Remove deprecated file types from interface (got rid of `std::old_io`)
* Introducing `FileType`

## 0.2.4 - 2015-02-22

* Update to latest Rust nightly
* `spawn_mount` returns a `Result` now since starting a new thread may fail
* Paths are now passed using `std::path::Path` (got rid of `std::old_path`)

native.rs
// Copyright 2013 Andreas Neuhaus and the Mira project group.
// This file may not be used in any way (copied, modified, distributed)
// without explicit permission of the copyright owner. All rights reserved.

/*!
 * FUSE native interface declarations (as of libosxfuse-2.5.5).
 */

#[allow(missing_doc)];

use std::libc::{c_int, c_char};

// Link with libosxfuse on OS X
#[cfg(target_os = "macos")]
#[link_args = "-losxfuse"]
extern {}

// Link with libfuse on Linux
#[cfg(target_os = "linux")]
#[link_args = "-lfuse"]
extern {}

/*
 * FUSE arguments (see fuse_opt.h for details)
 */

pub struct fuse_args {
	argc: c_int,
	argv: **c_char,
	allocated: c_int,
}

/*
 * FUSE common (see fuse_common_compat.h for details)
 */

extern {
	pub fn fuse_mount_compat25 (mountpoint: *c_char, args: *fuse_args) -> c_int;
	pub fn fuse_unmount_compat22 (mountpoint: *c_char);
}

/*
 * FUSE kernel (see fuse_kernel.h for details)
 */

pub static FUSE_KERNEL_VERSION: u32 = 7;
pub static FUSE_KERNEL_MINOR_VERSION: u32 = 8;
pub static FUSE_ROOT_ID: u64 = 1;

#[cfg(target_os = "macos")]
pub struct fuse_attr {
	ino: u64,
	size: u64,
	blocks: u64,
	atime: i64,
	mtime: i64,
	ctime: i64,
	crtime: i64,			// OS X only
	atimensec: i32,
	mtimensec: i32,
	ctimensec: i32,
	crtimensec: i32,		// OS X only
	mode: u32,
	nlink: u32,
	uid: u32,
	gid: u32,
	rdev: u32,
	flags: u32,				// OS X only, see chflags(2)
}

#[cfg(target_os = "linux")]
pub struct fuse_attr {
	ino: u64,
	size: u64,
	blocks: u64,
	atime: i64,
	mtime: i64,
	ctime: i64,
	atimensec: i32,
	mtimensec: i32,
	ctimensec: i32,
	mode: u32,
	nlink: u32,
	uid: u32,
	gid: u32,
	rdev: u32,
}

pub struct fuse_kstatfs {
	blocks: u64,			// Total blocks (in units of frsize)
	bfree: u64,				// Free blocks
	bavail: u64,			// Free blocks for unprivileged users
	files: u64,				// Total inodes
	ffree: u64,				// Free inodes
	bsize: u32,				// Filesystem block size
	namelen: u32,			// Maximum filename length
	frsize: u32,			// Fundamental file system block size
	padding: u32,
	spare: [u32, ..6],
}

pub struct fuse_file_lock {
	start: u64,
	end: u64,
	typ: u32,
	pid: u32,
}

pub mod consts {
	// Bitmasks for fuse_setattr_in.valid
	pub static FATTR_MODE: u32				= 1 << 0;
	pub static FATTR_UID: u32				= 1 << 1;
	pub static FATTR_GID: u32				= 1 << 2;
	pub static FATTR_SIZE: u32				= 1 << 3;
	pub static FATTR_ATIME: u32				= 1 << 4;
	pub static FATTR_MTIME: u32				= 1 << 5;
	pub static FATTR_FH: u32				= 1 << 6;
	#[cfg(target_os = "macos")]
	pub static FATTR_CRTIME: u32			= 1 << 28;	// OS X only
	#[cfg(target_os = "macos")]
	pub static FATTR_CHGTIME: u32			= 1 << 29;	// OS X only
	#[cfg(target_os = "macos")]
	pub static FATTR_BKUPTIME: u32			= 1 << 30;	// OS X only
	#[cfg(target_os = "macos")]
	pub static FATTR_FLAGS: u32				= 1 << 31;	// OS X only

	// Flags returned by the open request
	pub static FOPEN_DIRECT_IO: u32			= 1 << 0;	// bypass page cache for this open file
	pub static FOPEN_KEEP_CACHE: u32		= 1 << 1;	// don't invalidate the data cache on open
	#[cfg(target_os = "macos")]
	pub static FOPEN_PURGE_ATTR: u32		= 1 << 30;	// OS X only
	#[cfg(target_os = "macos")]
	pub static FOPEN_PURGE_UBC: u32			= 1 << 31;	// OS X only

	// Init request/reply flags
	pub static FUSE_ASYNC_READ: u32			= 1 << 0;
	pub static FUSE_POSIX_LOCKS: u32		= 1 << 1;
	#[cfg(target_os = "macos")]
	pub static FUSE_CASE_INSENSITIVE: u32	= 1 << 29;	// OS X only
	#[cfg(target_os = "macos")]
	pub static FUSE_VOL_RENAME: u32			= 1 << 30;	// OS X only
	#[cfg(target_os = "macos")]
	pub static FUSE_XTIMES: u32				= 1 << 31;	// OS X only

	// Release flags
	pub static FUSE_RELEASE_FLUSH: u32		= 1 << 0;
}

pub enum fuse_opcode {
	FUSE_LOOKUP = 1,
	FUSE_FORGET = 2,		// no reply
	FUSE_GETATTR = 3,
	FUSE_SETATTR = 4,
	FUSE_READLINK = 5,
	FUSE_SYMLINK = 6,
	FUSE_MKNOD = 8,
	FUSE_MKDIR = 9,
	FUSE_UNLINK = 10,
	FUSE_RMDIR = 11,
	FUSE_RENAME = 12,
	FUSE_LINK = 13,
	FUSE_OPEN = 14,
	FUSE_READ = 15,
	FUSE_WRITE = 16,
	FUSE_STATFS = 17,
	FUSE_RELEASE = 18,
	FUSE_FSYNC = 20,
	FUSE_SETXATTR = 21,
	FUSE_GETXATTR = 22,
	FUSE_LISTXATTR = 23,
	FUSE_REMOVEXATTR = 24,
	FUSE_FLUSH = 25,
	FUSE_INIT = 26,
	FUSE_OPENDIR = 27,
	FUSE_READDIR = 28,
	FUSE_RELEASEDIR = 29,
	FUSE_FSYNCDIR = 30,
	FUSE_GETLK = 31,
	FUSE_SETLK = 32,
	FUSE_SETLKW = 33,
	FUSE_ACCESS = 34,
	FUSE_CREATE = 35,
	FUSE_INTERRUPT = 36,
	FUSE_BMAP = 37,
	FUSE_DESTROY = 38,
	FUSE_SETVOLNAME = 61,	// OS X only
	FUSE_GETXTIMES = 62,	// OS X only
	FUSE_EXCHANGE = 63,		// OS X only
}

pub struct fuse_entry_out {
	nodeid: u64,
	generation: u64,
	entry_valid: i64,
	attr_valid: i64,
	entry_valid_nsec: i32,
	attr_valid_nsec: i32,
	attr: fuse_attr,
}

pub struct fuse_forget_in {
	nlookup: u64,
}

pub struct fuse_attr_out {
	attr_valid: i64,
	attr_valid_nsec: i32,
	dummy: u32,
	attr: fuse_attr,
}

#[cfg(target_os = "macos")]
pub struct fuse_getxtimes_out {	// OS X only
	bkuptime: i64,
	crtime: i64,
	bkuptimensec: i32,
	crtimensec: i32,
}

pub struct fuse_mknod_in {
	mode: u32,
	rdev: u32,
}

pub struct fuse_mkdir_in {
	mode: u32,
	padding: u32,
}

pub struct fuse_rename_in {
	newdir: u64,
}

#[cfg(target_os = "macos")]
pub struct fuse_exchange_in {	// OS X only
	olddir: u64,
	newdir: u64,
	options: u64,
}

pub struct fuse_link_in {
	oldnodeid: u64,
}

#[cfg(target_os = "macos")]
pub struct fuse_setattr_in {
	valid: u32,
	padding: u32,
	fh: u64,
	size: u64,
	unused1: u64,
	atime: i64,
	mtime: i64,
	unused2: u64,
	atimensec: i32,
	mtimensec: i32,
	unused3: u32,
	mode: u32,
	unused4: u32,
	uid: u32,
	gid: u32,
	unused5: u32,
	bkuptime: i64,			// OS X only
	chgtime: i64,			// OS X only
	crtime: i64,			// OS X only
	bkuptimensec: i32,		// OS X only
	chgtimensec: i32,		// OS X only
	crtimensec: i32,		// OS X only
	flags: u32,				// OS X only
}

#[cfg(target_os = "linux")]
pub struct fuse_setattr_in {
	valid: u32,
	padding: u32,
	fh: u64,
	size: u64,
	unused1: u64,
	atime: i64,
	mtime: i64,
	unused2: u64,
	atimensec: i32,
	mtimensec: i32,
	unused3: u32,
	mode: u32,
	unused4: u32,
	uid: u32,
	gid: u32,
	unused5: u32,
}

pub struct fuse_open_in {
	flags: u32,
	mode: u32,
}

pub struct fuse_open_out {
	fh: u64,
	open_flags: u32,
	padding: u32,
}

pub struct fuse_release_in {
	fh: u64,
	flags: u32,
	release_flags: u32,
	lock_owner: u64,
}

pub struct fuse_flush_in {
	fh: u64,
	unused: u32,
	padding: u32,
	lock_owner: u64,
}

pub struct fuse_read_in {
	fh: u64,
	offset: u64,
	size: u32,
	padding: u32,
}

pub struct fuse_write_in {
	fh: u64,
	offset: u64,
	size: u32,
	write_flags: u32,
}

pub struct fuse_write_out {
	size: u32,
	padding: u32,
}

pub struct fuse_statfs_out {
	st: fuse_kstatfs,
}

pub struct fuse_fsync_in {
	fh: u64,
	fsync_flags: u32,
	padding: u32,
}

#[cfg(target_os = "macos")]
pub struct fuse_setxattr_in {
	size: u32,
	flags: u32,
	position: u32,			// OS X only
	padding: u32,			// OS X only
}

#[cfg(target_os = "linux")]
pub struct fuse_setxattr_in {
	size: u32,
	flags: u32,
}

#[cfg(target_os = "macos")]
pub struct fuse_getxattr_in {
	size: u32,
	padding: u32,
	position: u32,			// OS X only
	padding2: u32,			// OS X only
}

#[cfg(target_os = "linux")]
pub struct fuse_getxattr_in {
	size: u32,
	padding: u32,
}

pub struct fuse_getxattr_out {
	size: u32,
	padding: u32,
}

pub struct fuse_lk_in {
	fh: u64,
	owner: u64,
	lk: fuse_file_lock,
}

pub struct fuse_lk_out {
	lk: fuse_file_lock,
}

pub struct fuse_access_in {
	mask: u32,
	padding: u32,
}

pub struct fuse_init_in {
	major: u32,
	minor: u32,
	max_readahead: u32,
	flags: u32,
}

pub struct fuse_init_out {
	major: u32,
	minor: u32,
	max_readahead: u32,
	flags: u32,
	unused: u32,
	max_write: u32,
}

pub struct fuse_interrupt_in {
	unique: u64,
}

pub struct fuse_bmap_in {
	block: u64,
	blocksize: u32,
	padding: u32,
}

pub struct fuse_bmap_out {
	block: u64,
}

pub struct fuse_in_header {
	len: u32,
	opcode: u32,
	unique: u64,
	nodeid: u64,
	uid: u32,
	gid: u32,
	pid: u32,
	padding: u32,
}

pub struct fuse_out_header {
	len: u32,
	error: i32,
	unique: u64,
}

pub struct fuse_dirent {
	ino: u64,
	off: u64,
	namelen: u32,
	typ: u32,
	// followed by name of namelen bytes
}
